File: app/src/main/java/com/kuplay/kuplay/common/js/JSBridge.java
package com.kuplay.kuplay.common.js;

import android.app.Activity;
import android.util.Log;
import android.webkit.JavascriptInterface;

import com.kuplay.kuplay.common.JSExecutable;
import com.kuplay.kuplay.util.CodeUtil;
import com.kuplay.kuplay.util.ContainerUtil;
import com.kuplay.kuplay.util.Logger;

import org.json.JSONArray;

import java.util.List;

/**
 * Created by "iqos_jay@outlook.com" on 2018/6/27.
 * Android and js bridging files for js call the local underlying method.
 */
public class JSBridge {
    private static final String TAG = JSBridge.class.getName();
    private static final String METHOD_INIT = "init";//method name->init
    private static final String METHOD_CLOSE = "close";//method name->close
    private final Object mWebView;
    private final Activity mActivity;

    /**
     * Constructor used to initialize classes that would be called by JS.
     * Get all classes from the project which implemented the {@link JSExecutable},
     * and put these classes into the JSBridge.
     * The file is needn't modify anymore,while you want add new features,
     * just make the class implement {@link JSExecutable}.
     */
    public JSBridge(Object webView, Activity activity) {
        this.mActivity = activity;
        this.mWebView = webView;
        List<Class> classes = CodeUtil.getAllClassByInterface(JSExecutable.class);
        if (ContainerUtil.isNullOrEmpty(classes)) return;
        for (Class clazz : classes) {
            JSEnv.setClass(clazz.getSimpleName(), clazz);
        }
    }

    /**
     * Called when js web wanna to call local function,
     * this class isn't created by me,other parameter will be filled after a while.
     *
     * @param className  Class's name that was called.
     * @param methodName Class's method that was called.
     */
    @JavascriptInterface
    public void postMessage(String className, String methodName, int nativeID, int listenerID, String jsonArray) {
        try {
            JSONArray js = new JSONArray(jsonArray);
            Object[] params = new Object[1 + js.length()];
            params[0] = listenerID;
            for (int i = 0; i < js.length(); i++) {
                params[i + 1] = js.get(i);
            }
            Log.d("JSBridge", "postMessage (" + className + ", " + methodName + ", " + nativeID + ", " + listenerID + ", " + jsonArray + ")");
            switch (methodName) {
                case METHOD_INIT:
                    int id = JSEnv.newInstance(className, mWebView, mActivity);
                    JSCallback.callJS(mActivity, listenerID, JSCallback.SUCCESS, id);
                    break;
                case METHOD_CLOSE:
                    JSEnv.removeObject(nativeID);
                    JSCallback.callJS(mActivity, listenerID, JSCallback.SUCCESS);
                    break;
                default:
                    JSEnv.call(className, methodName, nativeID, params);
                    break;
            }
        } catch (Exception e) {
            JSCallback.throwJS(mActivity, className, methodName, e.getMessage());
        }
    }
}

File: app/src/main/java/com/kuplay/kuplay/common/js/JSBridge.kt
package com.kuplay.kuplay.common.js

import android.app.Activity
import android.util.Log
import android.webkit.JavascriptInterface

import com.kuplay.kuplay.common.JSExecutable
import com.kuplay.kuplay.util.CodeUtil
import com.kuplay.kuplay.util.ContainerUtil

import org.json.JSONArray

/**
 * Created by "iqos_jay@outlook.com" on 2018/6/27.
 * Android and js bridging files for js call the local underlying method.
 */
class JSBridge
/**
 * Constructor used to initialize classes that would be called by JS.
 * Get all classes from the project which implemented the [JSExecutable],
 * and put these classes into the JSBridge.
 * The file is needn't modify anymore,while you want add new features,
 * just make the class implement [JSExecutable].
 */
(private val mWebView: Any, private val mActivity: Activity) {

    init {
        val classes = CodeUtil.getAllClassByInterface(JSExecutable::class.java)
        if (!ContainerUtil.isNullOrEmpty(classes)){
            for (clazz in classes) {
                JSEnv.setClass(clazz.simpleName, clazz)
            }
        }

    }

    /**
     * Called when js web wanna to call local function,
     * this class isn't created by me,other parameter will be filled after a while.
     *
     * @param className  Class's name that was called.
     * @param methodName Class's method that was called.
     */
    @JavascriptInterface
    fun postMessage(className: String, methodName: String, nativeID: Int, listenerID: Int, jsonArray: String) {
        Log.d("JSBridge","$className,$methodName,$jsonArray")
        var callBack = { callType: Int, prames: Array<Any> ->
            if (callType == JSCallback.SUCCESS){
                JSCallback.callJS(mActivity, mWebView, listenerID, callType, prames)
            }else if (callType == JSCallback.FAIL){
                JSCallback.throwJS(mActivity, mWebView, className, methodName, prames[0].toString())
            }
        }
        try {
            val js = JSONArray(jsonArray)
            val params = arrayOfNulls<Any>(1 + js.length())
            for (i in 0 until params.size) {
                if (i == js.length()){
                    params[i] = callBack
                }
                else {
                    params[i] = js.get(i)
                }
            }
            Log.d("JSEnv.call000","${params.size}")

            for (o in params){
                Log.d("JSEnv.call333","$o")
            }
            when (methodName) {
                METHOD_INIT -> {
                    val id = JSEnv.newInstance(className, mWebView, mActivity)
                    JSCallback.callJS(mActivity, mWebView, listenerID, JSCallback.SUCCESS, arrayOf(id))
                }
                METHOD_CLOSE -> {
                    JSEnv.removeObject(nativeID)
                    JSCallback.callJS(mActivity, mWebView, listenerID, JSCallback.SUCCESS, emptyArray())
                }
                else -> JSEnv.call(className, methodName, nativeID, params)

            }
        } catch (e: Exception) {
            JSCallback.throwJS(mActivity, mWebView, className, methodName, e.message!!)
        }

    }

    companion object {
        private val TAG = JSBridge::class.java.name
        private val METHOD_INIT = "init"//method name->init
        private val METHOD_CLOSE = "close"//method name->close
    }
}
